feat: Improve RAG injector documentation, add a real-world scenario how-to (#1522)
Kong/developer.konghq.com: app/_how-tos/use-ai-rag-injector-plugin.md

                temperature: 1.0
variables:
  openai_api_key:
    value: $OPENAI_API_KEY
{% endentity_examples %}

## Configure the AI RAG Injector plugin

Then, configure the AI RAG Injector plugin

{% entity_examples %}
entities:
  plugins:
  - name: ai-rag-injector
    config:
      inject_template: |
        You are an AI assistant designed to answer employee questions using only the approved compliance content provided between the <RAG></RAG> tags.
        Do not use external or general knowledge, and do not answer if the information is not available in the RAG content.
        <RAG><CONTEXT></RAG>
        User's question: <PROMPT>
        Respond only with information found in the <RAG> section. If the answer is not clearly present, reply with:
        "I'm sorry, I cannot answer that based on the available compliance information."
      embeddings:
        auth:
          header_name: Authorization
          header_value: Bearer ${openai_key}
        model:
          provider: openai
          name: text-embedding-3-large
      vectordb:
        strategy: redis
        redis:
          host: ${redis_host}
          port: 6379
        distance_metric: cosine

>         dimensions: 76

Reviewer: ```suggestion
        dimensions: 3072
```
this may not be correct, to match text-embedding-3-large this would be 3072

variables:
  openai_api_key:
    value: $OPENAI_API_KEY
  redis_host:
    value: $REDIS_HOST
{% endentity_examples %}

## Split input data before ingestion

Before sending data to the AI Gateway, split your input into manageable chunks using a text splitting tool like `langchain_text_splitters`. This helps optimize downstream processing and improves semantic retrieval performance.

> The following Python script demonstrates how to split text using `RecursiveCharacterTextSplitter` and ingest the resulting chunks into the AI Gateway:

Reviewer: ```suggestion
Please refer to [langchain text_splitters documents](https://python.langchain.com/docs/concepts/text_splitters/) if your documents
are structured data other than plain texts.

The following Python script demonstrates how to split text using `RecursiveCharacterTextSplitter` and ingest the resulting chunks into the AI Gateway:
```

```bash
curl localhost:8001/ai-rag-injector/3194f12e-60c9-4cb6-9cbc-c8fd7a00cff1/ingest_chunk \
  -H "Content-Type: application/json" \
  -d '{
    "content": "<chunk>"
  }'
```

## Make an AI request to the AI Proxy Advanced plugin

Once vector database has ingested data and built a knowledge base, you can make requests to it.
For example:

```bash
curl  --http1.1 localhost:8000/chat \
  -H "Content-Type: application/json" \
  -H "Authorization: Bearer $OPENAI_API_KEY" \
  -d '{
     "messages": [{"role": "user", "content": "What is kong"}]
   }' | jq
```

### Debug the retrieval of the knowledge base

To evaluate which documents are retrieved for a specific prompt, use the following command:

```bash
curl localhost:8001/ai-rag-injector/3194f12e-60c9-4cb6-9cbc-c8fd7a00cff1/lookup_chunks \
  -H "Content-Type: application/json" \
  -d '{
    "prompt": "the prompt to debug",
    "exclude_contents": false
  }'
```

{:.info}
> To omit the chunk content and only return the chunk ID, set `exclude_contents` to true.

> ## Update content for ingesting

Reviewer: we might want to re-order this section a bit (to before 'Ingest content to the vector database") , as we mention user need to onboard Konnect in the docs but the python script is calling admin API which won't in konnect.

Author: I removed the Konnect prerequisite. Will that work?

tags:
  - ai
  - openai

tldr:
  q: How do I use the AI RAG Injector plugin to safeguard that my company chatbot responds with relevant questions regarding compliance policies?
  a: Use the AI RAG Injector plugin to integrate your company’s compliance policy documents as retrieval-augmented knowledge. Configure the plugin to inject context from these documents into chatbot prompts, ensuring it can generate relevant, accurate compliance-related questions dynamically during conversations.

tools:
  - deck

prereqs:
  inline:
    - title: OpenAI
      include_content: prereqs/openai
      icon_url: /assets/icons/openai.svg
    - title: Redis stack
      content: |
          To complete this task, you must have a Redis stack configured in your environment. Check [Redis website](https://redis.io/docs/latest/) to learn more.
      icon_url: /assets/icons/redis.svg
    - title: PgVector (optionally)
      content: |
        Test
      icon_url: /assets/icons/database.svg
    - title: Langchain splitters
      content: |
        To complete this tutorial, you'll need **Python (version 3.7 or later)** and `pip` installed on your machine. You can verify it by running:

        ```bash

>         python --version

Reviewer 2: ```suggestion
        python3
```

>         pip --version

Reviewer 2: ```suggestion
        python3 -m pip --version
```

         ```

        Once that's set up, install the required packages by running the following command in your terminal:
        ```

>         pip install langchain langchain_text_splitters requests

Reviewer 2: ```suggestion
        python3 -m pip install langchain langchain_text_splitters requests
```
The pip install commands weren't working for me even though I added pip to my path.

        ```
      icon_url: /assets/icons/python.svg
  entities:
    services:
        - example-service
    routes:
        - example-route

cleanup:
  inline:
    - title: Clean up Konnect environment
      include_content: cleanup/platform/konnect
      icon_url: /assets/icons/gateway.svg
    - title: Destroy the {{site.base_gateway}} container
      include_content: cleanup/products/gateway
      icon_url: /assets/icons/gateway.svg
---

## Configure the AI Proxy Advanced plugin

> First, you'll need to configure the AI Proxy Advanced plugin:

Reviewer 2: Include a sentence about why we have to configure this first.

              header_name: Authorization
              header_value: Bearer ${openai_api_key}
            model:
              provider: openai
              name: gpt-4o
              options:
                max_tokens: 512
                temperature: 1.0
variables:
  openai_api_key:
    value: $OPENAI_API_KEY
{% endentity_examples %}

## Configure the AI RAG Injector plugin

Then, configure the AI RAG Injector plugin

{% entity_examples %}
entities:
  plugins:
  - name: ai-rag-injector
    id: 3194f12e-60c9-4cb6-9cbc-c8fd7a00cff1
    config:
      inject_template: |
        You are an AI assistant designed to answer employee questions using only the approved compliance content provided between the <RAG></RAG> tags.
        Do not use external or general knowledge, and do not answer if the information is not available in the RAG content.
        <RAG><CONTEXT></RAG>
        User's question: <PROMPT>
        Respond only with information found in the <RAG> section. If the answer is not clearly present, reply with:
        "I'm sorry, I cannot answer that based on the available compliance information."
      embeddings:
        auth:
          header_name: Authorization

>           header_value: Bearer ${openai_key}

Reviewer 2: ```suggestion
          header_value: Bearer ${openai_api_key}
```

        model:
          provider: openai
          name: text-embedding-3-large
      vectordb:
        strategy: redis
        redis:
          host: ${redis_host}
          port: 6379
        distance_metric: cosine
        dimensions: 3072
variables:
  openai_api_key:
    value: $OPENAI_API_KEY
  redis_host:
    value: $REDIS_HOST

> {% endentity_examples %}

Reviewer 2: This is giving me the following error, both before and after I corrected the open api key variable:

```
zsh: parse error near `\n'
```

Author: I'll look into that

Author: I've corrected and re-tested the snippet: it should work now

{"metadata":{"ingest_duration":1286,"embeddings_tokens_count":141,"chunk_id":"c3d4e5f6-7890-1234-cd56-7890abcdef12"}}
{"metadata":{"ingest_duration":2892,"embeddings_tokens_count":168,"chunk_id":"d4e5f678-9012-3456-de78-90abcdef1234"}}
```

### Ingest content to the vector database

Now, you can feed the split chunks into AI Gateway using the Kong Admin API.

The following example shows how to ingest content to the vector database for building the knowledge base. The AI RAG Injector plugin uses the OpenAI `text-embedding-3-large` model to generate embeddings for the content and stores them in Redis.

```bash
curl localhost:8001/ai-rag-injector/{plugin_ID}/ingest_chunk \
  -H "Content-Type: application/json" \
  -d '{
    "content": "<chunk>"
  }'
```
{:.info}
Replace the `plugin_ID` placeholder with your AI RAG Injector plugin ID.

## Test RAG configuration

### In-scope questions

Use the following in-scope questions to verify that the AI responds accurately based on the approved compliance content and does not rely on external knowledge.

{% navtabs "In scope" %}
{% navtab "Basic questions" %}

  Use simple user questions that map directly to travel policy clauses:

>   ```bash
>   curl --http1.1 localhost:8000/chat \
>     -H "Content-Type: application/json" \
>     -H "Authorization: Bearer $OPENAI_API_KEY" \
>     -d '{
>       "messages": [{"role": "user", "content": "Are alcoholic beverages reimbursable?"}]
>     }' | jq
>   ````

Reviewer 2: ````suggestion
  {% validation request-check %}
  url: /anything
  headers:
    - 'Content-Type: application/json'
    - 'Authorization: Bearer $DECK_OPENAI_API_KEY'
  body:
    messages:
      - role: user
        content: Are alcoholic beverages reimbursable?
  {% endvalidation %}
````
This request-check format is what we should be using here. A few things I noted/changed:
- `/chat` isn't a valid path for our route, so I changed it to `/anything`, which is the path for the example_route we set in the prereqs
- we set `$DECK_OPENAI_API_KEY` in the prereq instead of the env var you have listed. I think the decK env var should also work here, though not the most ideal since it's a decK env var instead of regular.

  You can also ask this question:

>   ```bash
>   curl --http1.1 localhost:8000/chat \
>     -H "Content-Type: application/json" \
>     -H "Authorization: Bearer $OPENAI_API_KEY" \
>     -d '{
>       "messages": [{"role": "user", "content": "What documentation is required for travel reimbursement?"}]
>     }' | jq
>   ```

Reviewer 2: ````suggestion
  {% validation request-check %}
  url: /anything
  headers:
    - 'Content-Type: application/json'
    - 'Authorization: Bearer $DECK_OPENAI_API_KEY'
  body:
    messages:
      - role: user
        content: What documentation is required for travel reimbursement?
  {% endvalidation %}
````
Just be sure to double check my indentation and how this renders in the output :)

{% endnavtab %}
{% navtab "Intermediate questions" %}

  Use slightly more complex prompts involving multi-step policy logic or multiple clauses:

>   ```bash
>   curl --http1.1 localhost:8000/chat \
>     -H "Content-Type: application/json" \
>     -H "Authorization: Bearer $OPENAI_API_KEY" \
>     -d '{
>       "messages": [{"role": "user", "content": "Can I get reimbursed for internet charges during a business trip?"}]
>     }' | jq
>   ```

Reviewer 2: ````suggestion
{% validation request-check %}
url: /anything
headers:
  - 'Content-Type: application/json'
  - 'Authorization: Bearer $DECK_OPENAI_API_KEY'
body:
  messages:
    - role: user
      content: Can I get reimbursed for internet charges during a business trip?
{% endvalidation %}
````

  Also, you can ask a more complex query about booking a hotel:

>   ```bash
>   curl --http1.1 localhost:8000/chat \
>     -H "Content-Type: application/json" \
>     -d '{
>       "messages": [{"role": "user", "content": "Do I need to book my hotel in advance for business travel?"}]
>     }' | jq
>   ```

Reviewer 2: ```suggestion
{% validation request-check %}
url: /anything
headers:
  - 'Content-Type: application/json'
  - 'Authorization: Bearer $DECK_OPENAI_API_KEY'
body:
  messages:
    - role: user
      content: Do I need to book my hotel in advance for business travel?
{% endvalidation %}
```

{% endnavtab %}
{% navtab "Edge cases" %}

  Use prompts that test boundaries of the compliance language:

>   ```bash
>   curl --http1.1 localhost:8000/chat \
>     -H "Content-Type: application/json" \
>     -H "Authorization: Bearer $OPENAI_API_KEY" \
>     -d '{
>       "messages": [{"role": "user", "content": "Am I allowed to share a hotel room with another employee?"}]
>     }' | jq
>   ```

Reviewer 2: ```suggestion
{% validation request-check %}
url: /anything
headers:
  - 'Content-Type: application/json'
  - 'Authorization: Bearer $DECK_OPENAI_API_KEY'
body:
  messages:
    - role: user
      content: Am I allowed to share a hotel room with another employee?
{% endvalidation %}
```

  Or ask about public transportation:

>   ```bash
>   curl --http1.1 localhost:8000/chat \
>     -H "Content-Type: application/json" \
>     -H "Authorization: Bearer $OPENAI_API_KEY" \
>     -d '{
>       "messages": [{"role": "user", "content": "What’s the policy on using public transportation during travel?"}]
>     }' | jq
>   ```

Reviewer 2: ```suggestion
{% validation request-check %}
url: /anything
headers:
  - 'Content-Type: application/json'
  - 'Authorization: Bearer $DECK_OPENAI_API_KEY'
body:
  messages:
    - role: user
      content: What’s the policy on using public transportation during travel?
{% endvalidation %}
```

{% endnavtab %}
{% endnavtabs %}

### Out-of-scope questions

Use the following out-of-scope questions to confirm that the AI correctly refuses to answer queries that fall outside the ingested compliance content. Upon these requests, AI should return the following response:

```json
"message": {
    "role": "assistant",
    "content": "I'm sorry, I cannot answer that based on the available compliance information.",
  }
```

{% navtabs "test" %}
{% navtab "General company info" %}

  These questions ask about Acme Corp. in general, not about the travel policy:

>   ```bash
>   curl --http1.1 localhost:8000/chat \
>     -H "Content-Type: application/json" \
>     -H "Authorization: Bearer $OPENAI_API_KEY" \
>     -d '{
>       "messages": [{"role": "user", "content": "What does Acme Corp. do?"}]
>     }' | jq
>   ````

Reviewer 2: ```suggestion
{% validation request-check %}
url: /anything
headers:
  - 'Content-Type: application/json'
  - 'Authorization: Bearer $DECK_OPENAI_API_KEY'
body:
  messages:
    - role: user
      content: What does Acme Corp. do?
{% endvalidation %}
```

>   ```bash
>   curl --http1.1 localhost:8000/chat \
>     -H "Content-Type: application/json" \
>     -H "Authorization: Bearer $OPENAI_API_KEY" \
>     -d '{
>       "messages": [{"role": "user", "content": "Where is Acme Corp. headquartered?"}]
>     }' | jq
>   ```

Reviewer 2: ```suggestion
{% validation request-check %}
url: /anything
headers:
  - 'Content-Type: application/json'
  - 'Authorization: Bearer $DECK_OPENAI_API_KEY'
body:
  messages:
    - role: user
      content: Where is Acme Corp. headquartered?
{% endvalidation %}
```

{% endnavtab %}
{% navtab "External knowledge" %}

  These questions require general or external knowledge that is not included in the ingested content:

>   ```bash
>   curl --http1.1 localhost:8000/chat \
>     -H "Content-Type: application/json" \
>     -H "Authorization: Bearer $OPENAI_API_KEY" \
>     -d '{
>       "messages": [{"role": "user", "content": "Who is the CEO of OpenAI?"}]
>     }' | jq
>   ```
> 
>   ```bash
>   curl --http1.1 localhost:8000/chat \
>     -H "Content-Type: application/json" \
>     -H "Authorization: Bearer $OPENAI_API_KEY" \
>     -d '{
>       "messages": [{"role": "user", "content": "How does Redis handle vector storage?"}]
>     }' | jq
>   ```

Reviewer 2: ```suggestion
{% validation request-check %}
url: /anything
headers:
  - 'Content-Type: application/json'
  - 'Authorization: Bearer $DECK_OPENAI_API_KEY'
body:
  messages:
    - role: user
      content: Who is the CEO of OpenAI?
{% endvalidation %}

{% validation request-check %}
url: /anything
headers:
  - 'Content-Type: application/json'
  - 'Authorization: Bearer $DECK_OPENAI_API_KEY'
body:
  messages:
    - role: user
      content: How does Redis handle vector storage?
{% endvalidation %}
```

{% endnavtab %}
{% navtab "Other HR policies" %}

> These prompts reference company policies that are not part of the travel policy content:
> 
>   ```bash
>   curl --http1.1 localhost:8000/chat \
>     -H "Content-Type: application/json" \
>     -H "Authorization: Bearer $OPENAI_API_KEY" \
>     -d '{
>       "messages": [{"role": "user", "content": "How much vacation time do I get per year?"}]
>     }' | jq
>   ```
> 
>   ```bash
>   curl --http1.1 localhost:8000/chat \
>     -H "Content-Type: application/json" \
>     -H "Authorization: Bearer $OPENAI_API_KEY" \
>     -d '{
>       "messages": [{"role": "user", "content": "What’s the parental leave policy at Acme Corp.?"}]
>     }' | jq
>   ```

Reviewer 2: ```suggestion
These prompts reference company policies that aren't part of the travel policy content:

{% validation request-check %}
url: /anything
headers:
  - 'Content-Type: application/json'
  - 'Authorization: Bearer $DECK_OPENAI_API_KEY'
body:
  messages:
    - role: user
      content: How much vacation time do I get per year?
{% endvalidation %}

{% validation request-check %}
url: /anything
headers:
  - 'Content-Type: application/json'
  - 'Authorization: Bearer $DECK_OPENAI_API_KEY'
body:
  messages:
    - role: user
      content: What’s the parental leave policy at Acme Corp.?
{% endvalidation %}
```

{% endnavtab %}
{% navtab "Ambiguous or unsupported topics" %}

These prompts are vague, outside compliance scope, or might encourage hallucination if guardrails aren't working:

>   ```bash
>   curl --http1.1 localhost:8000/chat \
>     -H "Content-Type: application/json" \
>     -H "Authorization: Bearer $OPENAI_API_KEY" \
>     -d '{
>       "messages": [{"role": "user", "content": "What is the best destination for international travel?"}]
>     }' | jq
>   ```
> 
>   ```bash
>   curl --http1.1 localhost:8000/chat \
>     -H "Content-Type: application/json" \
>     -H "Authorization: Bearer $OPENAI_API_KEY" \
>     -d '{
>       "messages": [{"role": "user", "content": "What should I pack for an international trip?"}]
>     }' | jq
>   ```

Reviewer 2: ```suggestion
{% validation request-check %}
url: /anything
headers:
  - 'Content-Type: application/json'
  - 'Authorization: Bearer $DECK_OPENAI_API_KEY'
body:
  messages:
    - role: user
      content: What is the best destination for international travel?
{% endvalidation %}

{% validation request-check %}
url: /anything
headers:
  - 'Content-Type: application/json'
  - 'Authorization: Bearer $DECK_OPENAI_API_KEY'
body:
  messages:
    - role: user
      content: What should I pack for an international trip?
{% endvalidation %}
```

{% endnavtab %}
{% endnavtabs %}

### Debug the retrieval of the knowledge base

To evaluate which documents are retrieved for a specific prompt, use the following command:

> ```bash
> curl localhost:8001/ai-rag-injector/{plugin_ID}/lookup_chunks \
>   -H "Content-Type: application/json" \
>   -d '{
>     "prompt": "the prompt to debug",
>     "exclude_contents": false
>   }'
> ```

Reviewer 2: ```suggestion
<!--vale off-->
{% control_plane_request %}
url: /ai-rag-injector/b924e3e8-7893-4706-aacb-e75793a1d2e9/lookup_chunks
status_code: 201
headers:
    - 'Accept: application/json'
    - 'Content-Type: application/json'
body:
    prompt: Am I allowed to share a hotel room with another employee?
    exclude_contents: false
{% endcontrol_plane_request %}
<!--vale on-->

This will return which content in the compliance policy AI is using to answer the user question.
```
I think that last sentence is correct

     if #embeddings_vector ~= conf.vectordb.dimensions then
       return nil, "Embedding dimensions do not match the configured vector database. Embeddings were   " ..
         #embeddings_vector .. " dimensions, but the vector database is configured for " ..
         conf.vectordb.dimensions .. " dimensions.", "Embedding dimensions do not match the   configured vector database"
     end

     metadata.chunk_id = uuid()
     -- ingest chunk
     local _, err = vectordb_driver:insert(embeddings_vector, content, metadata.chunk_id)
     if err then
         return nil, "Failed to insert chunk: " .. err
     end

     return true
   end

   assert(#args == 3, "2 arguments expected")
   local plugin_id, content = args[2], args[3]

   local plugin, err = get_plugin_by_id(plugin_id)
   if err then
     ngx.log(ngx.ERR, "Failed to get plugin: " .. err)
     return
   end

   if not plugin then
     ngx.log(ngx.ERR, "Plugin not found")
     return
   end

   local _, err = ingest_chunk(plugin.config, content)
   if err then
     ngx.log(ngx.ERR, "Failed to ingest: " .. err)
     return
   end

   ngx.log(ngx.INFO, "Update completed")

   ```

> 3. Run the script from your Kong instance:
> 
>    ```sh
>    kong runner ingest_api.lua <plugin_id> <content_to_update>
>    ```

Reviewer 2: ````suggestion
3. Run the script from your Kong instance. This uses your AI RAG Injector plugin ID and the content you want to update. Here's an example:

   ```sh
   kong runner ingest_api.lua b924e3e8-7893-4706-aacb-e75793a1d2e9 ./inject_policy.py
   ```
````
I couldn't get this to work. I put the inject_policy.py file in there, but that didn't work. I think we need an example of the content to update because I'm not sure if it's a file or a specific line in the compliance policy or how it works.
